corehq/apps/accounting/static/accounting/js/confirm_plan.js

```diff
@@ -0,0 +1,79 @@
+hqDefine('accounting/js/confirm_plan', [
+    'jquery',
+    'knockout',
+    'underscore',
+    'hqwebapp/js/initial_page_data',
+], function (
+    $,
+    ko,
+    _,
+    initialPageData
+) {
+    var confirmPlanModel = function (isUpgrade, currentPlan, newPlan) {
+        'use strict';
+        var self = {};
+
+        self.isUpgrade = isUpgrade;
+        self.currentPlan = currentPlan;
+        self.newPlan = newPlan;
+
+        self.form = undefined;
+        self.openDowngradeModal = function(confirmPlanModel, e) {
+            self.form = $(e.currentTarget).closest("form");
+            if (confirmPlanModel.isUpgrade) {
+                self.form.submit();
+            } else {
+                var $modal = $("#modal-downgrade");
+                $modal.modal('show');
+            }
+        };
+        self.submitDowngrade = function(pricingTable, e) {
+            var finish = function() {
+                if (self.form) {
+                    self.form.submit();
+                }
+            };
+
+            var $button = $(e.currentTarget);
+            $button.disableButton();
+            $.ajax({
+                method: "POST",
+                url: initialPageData.reverse('email_on_downgrade'),
+                data: {
+                    old_plan: self.currentPlan.name,
+                    new_plan: self.newPlan.name,
+                    note: $button.closest(".modal").find("textarea").val(),
+                },
+                success: finish,
+                error: finish,
+            });
+        };
+
+        self.init = function () {
```

Comment:
Curious, is there a reason these states aren't in knockout? The code looks fine, it's just unusual to have `document.getElementById` and event handlers inside a knockout model.

Reply:
I'm still rather confused about how to properly use knockout. How would you connect the `user-agreement` and `confirm-plan` objects in HTML?

Reply:
There are a couple of changes you'd need to make:
1. Instead of calling `koApplyBindings` on `confirm-plan`, you'd call it on some ancestor element that contains both `confirm-plan` and `user-agreement`. In this case, it looks like `#hq-content` is the lowest ancestor that contains both the button and the checkbox, so you could use that. Usually `koApplyBindings` is called on some big container element, and then all of the HTML inside that container has access to the bindings\*.
1. In `confirmPlanModel`, you'd add an observable property to represent whether or not the user has signed the agreement. Something like `self.userAgreementSigned = ko.observable(false);`
1. In the HTML, you'd add a [checked](http://knockoutjs.com/documentation/checked-binding.html) binding to the checkbox that would link `userAgreementSigned` to the value of the checkbox: `data-bind="checked: userAgreementSigned"`. This would replace the `value="True"` that's there now (which btw I don't think is doing anything). If you want to test that this is done correctly, you can stick in a debug label in the HTML like this: `<span class="label label-info">userAgreementSigned is <span data-bind="text: userAgreementSigned"></span></span>` ...that'll make a little label show up that should display either "true" or "false" and update automatically as you check & uncheck the box.
1. In the HTML, you'd add an [enable](http://knockoutjs.com/documentation/enable-binding.html) binding to the button, which would cause the button to only be enabled when `userAgreementSigned` is true: `data-bind="enable: userAgreementSigned"`.

Those changes would be the knockout equivalent of the `init` function currently in `confirmPlanModel`.


\* Bonus fact: If you look at the [source for koApplyBindings](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/hq.helpers.js#L149-L158) you'll see it's a simple wrapper around knockout's built in `applyBindings` method. The reason we have `koApplyBindings` is because if you call `applyBindings` without passing it an element, it will apply the bindings to the entire page. We had bugs where someone would call `applyBindings` on an element that didn't exist (like something that's only on the page if you have some feature flag turned on) and then bindings would get applied to the whole page and cause errors. Anyway, the point of that is that knockout bindings generally apply to some fairly large unit, like a whole page.